intro_to_ada.rst

```diff
@@ -2681,21 +2681,1055 @@ Protected objects: entries
+           Convention    => C;
+
+       V : int;
+
+       func_cnt : int
+         with
+           Import        => True,
+           Convention    => C;
+       --  We can access the func_cnt variable from test.c
+
+    begin
+       V := my_func (1);
+       V := my_func (2);
+       V := my_func (3);
+       Put_Line ("Result is " & int'Image (V));
+
+       Put_Line ("Function was called " & int'Image (func_cnt) & " times");
+    end Show_C_Func;
+
+As we can see by running the application, the value from the counter will
+contain the correct number of times that ``my_func`` was called.
+
+Similar to subprograms, we could use the ``External_Name`` aspect to
+rename the variable in the Ada application.
+
+Using Ada variables in C
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+It is also possible to use variables declared in Ada files in C
+applications. Similarly to subprogram, this requires the use of the
+``Export`` aspect.
+
+Let's reuse the previous example and add a counter, as we did in the
+previous example:
+
+.. code-block:: ada
+
+    --% filename: c_api.ads
+
+    with Interfaces.C;
+    use  Interfaces.C;
+
+    package C_API is
+
+       func_cnt : int := 0
+         with
+           Export     => True,
+           Convention => C;
+
+       function My_Func (a : int) return int
+         with
+           Export        => True,
+           Convention    => C,
+           External_Name => "my_func";
+
+    end C_API;
+
+The variable is then increment in ``My_Func``:
+
+.. code-block:: ada
+
+    --% filename: c_api.adb
+    package body C_API is
+
+       function My_Func (a : int) return int is
+       begin
+          func_cnt := func_cnt + 1;
+          return a * 2;
+       end My_Func;
+
+    end C_API;
+
+In the C application, we just need to declare the variable and use it:
+
+.. code-block:: c
+
+    #include <stdio.h>
+
+    extern int my_func (int a);
+
+    extern int func_cnt;
+
+    int main (int argc, char **argv) {
+
+      int v;
+
+      v = my_func(1);
+      v = my_func(2);
+      v = my_func(3);
+
+      printf("Result is %d\n", v);
+
+      printf("Function was called %d times\n", func_cnt);
+
+      return 0;
+    }
+
+Again, by running the application, we see that the value from the counter
+will contain the correct number of times that ``my_func`` was called.
+
+Using unconstrained types
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In the previous examples, we're being careful about the data types: all of
+them are coming from the ``Interfaces.C`` package. Using Ada built-in
+types when interfacing with C can be problematic, especially in case of
+unconstrained types. For example:
+
+.. code-block:: c
+
+    /*% filename: test.h */
+
+    char * my_func (void);
+
+This is the function implementation:
+
+.. code-block:: c
+
+    #include <stdio.h>
+    #include "test.h"
+
+    char * my_func (void)
+    {
+      return "hello";
+    }
+
+In the Ada application, we try to import this as a ``String`` type:
```

Comment:
This paragraph is pretty unsatisfying :) Showing a way to do things that does not work is pretty bad. IMHO we should:

1. Say what does not work and why. No need to show a non working example that people might try to copy without reading the text.

2. Show in a later section how to actually interface those types.

Reply:
I don't think it's sufficient to mention what does not work without a source-code example. People should be able to "visualize" what exactly we're referring to. I was just trying to be very clear about an important pitfall when interfacing with C.

Also, people shouldn't just copy examples without reading, since this an introductory book, not an "Ada cookbook". If this was a cookbook, however, I'd completely agree with you!

As a compromise, I'm OK with moving this section on unconstrained types to the advanced course. Can we agree on that?

Reply:
> As a compromise, I'm OK with moving this section on unconstrained types to the advanced course. Can we agree on that?

I dont think this is what we are discussing here. I'm not saying this is too advanced, in fact I think it's exactly at the right level for this course. I'm saying this is unhelpful for the reader.

I think you should not make people read 25 lines of code that is wrong. You should not assume that people have read everything around it. The fact that "this is not an Ada cookbook" is not a good excuse. The fact that your code contains no comment about the fact that it is wrong makes it even worse because it's even easier to get confused.

As a reader I can read your first paragraph and forget that there is a problem, after all you're just saying "can be problematic". Then I'll piece out the code example, and then at the end I will be like "oh ok so I figured out all this, but it's not actually how it works ?"

> People should be able to "visualize"

But that's the point: Your example does not help me visualize what is wrong. It's just a long way of showing what does not work. Too long, since it does not work. Saying

> Careful! You cannot interface this C function
>
> ```c
> char * my_func (void) {
>   return "hello";
> }
> ```
>
> Via this Ada interface
>
> ```ada
> function my_func return String
>   with Import => True, Convention => C;
> ```
>
> Because [insert helpful explanation here]

Would be amply enough. Also you could focus on the actual explanation, which is missing from your write up:

> the recommendation is to be very careful about the data types
> and use the ``Interfaces.C`` package whenever possible for interfacing
> with C.

I read this like "this is black magic". Well this is not the case. Nothing about this is magic. The representation of data is just very different, because Ada arrays (and hence strings) carry their bounds around, and C strings do not. That's why your example does not work, and this is what you should explain, rather than showing a full erroneous example.

Telling people to use Interfaces.C whenever possible is good, but not good enough here.

Reply:
I guess we can summarize my feedback in two points:

1. You're missing an actual explanation of what does not work and why. This should be the main meat of your paragraph, and for the moment it is missing.
2. You're missing an example where you make it work. It should go there because it's not actually that complicated, using `Interfaces.C`.
3. You're pushing the "complete code examples" too far. Complete code examples are useful when the code is actually correct.

Reply:
I've already removed this subsection in the subsequent changes from last Friday. Since you have a very clear idea in your head about how this subsection should look like, would you mind writing that down (even if it's just a draft)? I'm glad to review that when you're finished.

Reply:
Ok no problem, I'll take care of a first draft of that part and then we can see from there. I'll merge your changes since this is then only point of discussion, rest looks good to me!